src/xdp/apikernel.c

```diff
@@ -483,15 +483,28 @@ XdpApiKernelStart(
         Characteristics->ProviderRegistrationInstance.Size = sizeof(NPI_REGISTRATION_INSTANCE);
         Characteristics->ProviderRegistrationInstance.NpiId = &NPI_XDPAPI_INTERFACE_ID;
         Characteristics->ProviderRegistrationInstance.ModuleId = &NPI_XDP_MODULEID;
-        Characteristics->ProviderRegistrationInstance.Number = XDP_API_VERSION_1;
-
-        Status = NmrRegisterProvider(
-                Characteristics,
-                &XdpApiProvider,
-                &XdpApiProvider.NmrProviderHandle);
-        if (!NT_SUCCESS(Status)) {
-            TraceError(TRACE_CORE, "NmrRegisterProvider failed Status=%!STATUS!", Status);
-            goto Exit;
+
+        // Assume all API versions are supported.
+        BYTE Version = XDP_API_VERSION_1;
+        int HandleIdx = 0;
+        for (; HandleIdx < ARRAYSIZE(XdpApiProvider.NmrProviderHandles); Version++, HandleIdx++) {
```

Comment:
style suggestion: it is probably simpler to write this loop in terms of `i` and maintain a table of unique provider characteristics, in including versions.